File: src/net/liplum/blocks/cloud/SharedRoom.java
package net.liplum.blocks.cloud;

import arc.struct.OrderedSet;
import arc.util.Reflect;
import arc.util.serialization.Json;
import arc.util.serialization.JsonValue;
import mindustry.Vars;
import mindustry.gen.Building;
import mindustry.world.modules.ItemModule;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class SharedRoom implements Json.JsonSerializable {
    private final transient OrderedSet<Building> users = new OrderedSet<>();
    @Nullable
    public ItemModule sharedItemModule;
    private transient boolean onlyOnCloud = false;
    @NotNull
    public transient CloudInfo sharedInfo = new CloudInfo();
    public transient Runnable whenOfflineLastOne = null;

    public void online(IShared sharedBuild) {
        if (sharedItemModule == null) {
            sharedItemModule = sharedBuild.getSharedItems();
        } else {
            sharedBuild.setSharedItems(sharedItemModule);
        }
        sharedBuild.setSharedInfo(sharedInfo);
        users.add(sharedBuild.getBuilding());
    }

    public void offline(IShared sharedBuild) {
        Building b = sharedBuild.getBuilding();
        if (users.contains(b)) {
            if (users.size == 1 && whenOfflineLastOne != null) {
                onlyOnCloud = true;
                whenOfflineLastOne.run();
            }
            users.remove(b);
        }
    }

    @Override
    public void write(Json json) {
        if (sharedItemModule != null) {
            int[] itemArray = Reflect.get(sharedItemModule, "items");
            json.writeValue("Items", itemArray);
            json.writeValue("Total", sharedItemModule.total());
        }
        json.writeValue("OnlyOnCloud", onlyOnCloud);
    }

    @Override
    public void read(Json json, JsonValue jsonValue) {
        JsonValue onlyOnCloudJV = jsonValue.get("OnlyOnCloud");
        if (onlyOnCloudJV != null) {
            onlyOnCloud = onlyOnCloudJV.asBoolean();
            if (onlyOnCloud) {
                sharedItemModule = new ItemModule();
                JsonValue itemsJV = jsonValue.get("Items");
                int[] items = new int[Vars.content.items().size];
                if (itemsJV != null) {
                    int[] formerItems = itemsJV.asIntArray();
                    System.arraycopy(formerItems, 0, items, 0, formerItems.length);
                    Reflect.set(sharedItemModule, "items", items);
                }
                JsonValue totalJV = jsonValue.get("Total");
                if (totalJV != null) {
                    Reflect.set(sharedItemModule, "total", totalJV.asInt());
                }
            }
        }
    }
}

File: src/net/liplum/blocks/cloud/SharedRoom.kt
package net.liplum.blocks.cloud

import arc.struct.Seq
import arc.util.io.Reads
import arc.util.io.Writes
import mindustry.gen.Building
import mindustry.world.modules.ItemModule
import net.liplum.persistance.IRWable

open class SharedRoom : IRWable {
    var usersPos: Seq<Int>
    var sharedItemModule: ItemModule? = null

    constructor() {
        usersPos = Seq()
    }

    constructor(usersPos: Seq<Int>) {
        this.usersPos = usersPos
    }

    open fun online(sharedBuild: IShared) {
        if (sharedItemModule == null) {
            sharedItemModule = sharedBuild.sharedItems
        } else {
            sharedBuild.sharedItems = sharedItemModule!!
        }
    }

    open fun checkExist(isExisted: (Int) -> Boolean) {
        val it = usersPos.iterator()
        while (it.hasNext()) {
            val cur = it.next()
            if (!isExisted(cur)) {
                it.remove()
            }
        }
    }

    override fun read(reader: Reads) {
        val len = reader.i()
        val seq = Seq<Int>(len)
        for (i in 0..len) {
            seq.add(i);
        }
        usersPos = seq
    }

    override fun write(writer: Writes) {
        writer.i(usersPos.size)
        for (pos in usersPos) {
            writer.i(pos)
        }
    }

    companion object {
        @JvmStatic
        fun <Owner : Building> readFrom(reader: Reads): SharedRoom {
            val len = reader.i()
            val seq = Seq<Int>(len)
            for (i in 0..len) {
                seq.add(i);
            }
            return SharedRoom(seq)
        }
    }
}